src/test_fixture/event_gen.rs

```diff
@@ -0,0 +1,299 @@
+#[derive(Copy, Clone, Hash, Ord, PartialOrd, Eq, PartialEq)]
+struct UserId(u64);
+
+impl From<u64> for UserId {
+    fn from(value: u64) -> Self {
+        Self(value)
+    }
+}
+
+impl From<UserId> for u64 {
+    fn from(value: UserId) -> Self {
+        value.0
+    }
+}
+
+impl From<UserId> for usize {
+    fn from(value: UserId) -> Self {
+        usize::try_from(value.0).unwrap()
+    }
+}
+
+#[derive(Debug)]
+#[cfg_attr(feature = "clap", derive(clap::Args))]
+pub struct Config {
+    #[cfg_attr(feature = "clap", arg(long, default_value = "1000000000000"))]
+    pub max_user_id: NonZeroU64,
+    #[cfg_attr(feature = "clap", arg(long, default_value = "5"))]
+    pub max_trigger_value: NonZeroU32,
+    #[cfg_attr(feature = "clap", arg(long, default_value = "20"))]
+    pub max_breakdown_key: NonZeroU32,
+    #[cfg_attr(feature = "clap", arg(long, default_value = "10"))]
+    pub max_events_per_user: NonZeroU32,
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        Self::new(1_000_000_000_000, 5, 20, 50)
+    }
+}
+
+impl Config {
+    /// Creates a new instance of [`Self`]
+    ///
+    /// ## Panics
+    /// If any argument is 0.
+    #[must_use]
+    pub fn new(
+        max_user_id: u64,
+        max_trigger_value: u32,
+        max_breakdown_key: u32,
+        max_events_per_user: u32,
+    ) -> Self {
+        Self {
+            max_user_id: NonZeroU64::try_from(max_user_id).unwrap(),
+            max_trigger_value: NonZeroU32::try_from(max_trigger_value).unwrap(),
+            max_breakdown_key: NonZeroU32::try_from(max_breakdown_key).unwrap(),
+            max_events_per_user: NonZeroU32::try_from(max_events_per_user).unwrap(),
+        }
+    }
+
+    fn max_user_id(&self) -> usize {
+        usize::try_from(self.max_user_id.get()).unwrap()
+    }
+}
+
+use crate::{rand::Rng, test_fixture::ipa::TestRawDataRecord};
+use std::{
+    collections::HashSet,
+    num::{NonZeroU32, NonZeroU64},
+};
+
+struct UserStats {
+    user_id: UserId,
+    generated: u32,
+    max: u32,
+}
+
+impl UserStats {
+    fn new(user_id: UserId, max_events: u32) -> Self {
+        Self {
+            user_id,
+            generated: 0,
+            max: max_events,
+        }
+    }
+
+    fn add_one(&mut self) -> bool {
+        debug_assert!(self.generated < self.max);
+
+        self.generated += 1;
+        self.generated == self.max
+    }
+}
+
+/// Generates random source and trigger events with guarantee that every next event
+/// occurs at the same time or after the previous event.
+///
+/// Number of events generated depends on the configured number of unique users per set
+/// and maximum number of events per user. See [`Config`] for more details
+///
+/// [`Config`]: Config
+pub struct EventGenerator<R: Rng> {
+    config: Config,
+    rng: R,
+    users: Vec<UserStats>,
+    // even bit vector takes too long to initialize for MAX_USER_ID. Need a sparse structure
+    // here
+    used: HashSet<UserId>,
```

Comment:
Do we even need to track this?  I mean, the problem here is that we're essentially stateless, but you could create user IDs in monotonically increasing sequence.  A PRP would allow you to have the identifiers that you use appear to be randomly distributed in the target range...